File: src/storage/models/episode.py
from sqlalchemy import Column, Integer, String, ForeignKey
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from storage.database import Base
from storage.models.character import episode_appearances


class Episode(Base):
    __tablename__ = 'episodes'

    id = Column(UUID(as_uuid=True), primary_key=True)
    anime_id = Column(UUID(as_uuid=True), ForeignKey('animes.id'))

    season_id = Column(UUID(as_uuid=True), ForeignKey('seasons.id'), nullable=True)

    # file download
    download_id = Column(UUID(as_uuid=True), ForeignKey('downloads.id'), nullable=True)

    # episode numbers aren't always integers
    # Example: OVA, SP1 etc...
    episode_number = Column(String)
    length = Column(Integer)

    sub_group = Column(String, default='__UNKNOWN__')

    # 2 letter country code [EN, JP]
    language = Column(String, default='EN')

    characters = relationship('Character', secondary=episode_appearances, back_populates='episodes')

    dialogues = relationship('Dialogue', back_populates='episode')

    anime = relationship('Anime', back_populates='episodes')

    season = relationship('Season', back_populates='episodes')

    download = relationship('Download', back_populates='episodes')

File: src/database/entities/episode.ts
/* tslint:disable */
import {
  AutoIncrement,
  BelongsTo, BelongsToMany,
  Column,
  CreatedAt,
  ForeignKey, HasMany,
  Model,
  PrimaryKey,
  Table, UpdatedAt
} from "sequelize-typescript";

import Anime from "./anime";
import Dialogue from "./dialogue";
import CharacterDiscovery from "./character_discovery";
import Download from "./download";

@Table({
  tableName: "episodes",
  underscored: true
})
export default class Episode extends Model<Episode> {
  @AutoIncrement @PrimaryKey @Column
  public readonly id: number;

  @Column({ comment: 'not always numeric' })
  public readonly episodeNumber: string;

  @Column
  public readonly name: string;

  @Column({ allowNull: false })
  public readonly length: number;

  @Column({ allowNull: false , defaultValue: 'EN' })
  public readonly language: string;

  @Column({ allowNull: false , defaultValue: '__UNKNOWN__' })
  public readonly subGroup: string;

  @Column
  public readonly thumbnailUrl: string;

  @Column({ allowNull: false }) @ForeignKey(() => Anime)
  public readonly animeId: number;

  @HasMany(() => CharacterDiscovery)
  public readonly characters: CharacterDiscovery[];

  @HasMany(() => Dialogue)
  public readonly dialogues: Dialogue[];

  @BelongsTo(() => Anime)
  public readonly anime: Anime;

  @Column @ForeignKey(() => Download)
  public readonly downloadId?: number;

  @BelongsTo(() => Download)
  public readonly download?: Download;

  @CreatedAt
  public readonly createdAt: Date;

  @UpdatedAt
  public readonly updatedAt: Date;
}
